glfs.c

```diff
@@ -510,15 +515,50 @@ static int tcmu_glfs_open(struct tcmu_device *dev)
 		goto unref;
 	}
 
-	return 0;
+	ret = glfs_lstat(gfsp->fs, gfsp->hosts->path, &st);
+	if (ret) {
+		tcmu_dev_warn(dev, "glfs_lstat failed: %m\n");
+		goto close;
+	}
+
+	dev_size = tcmu_get_dev_num_lbas(dev) * tcmu_get_dev_block_size(dev);
+	if (st.st_size != dev_size) {
+		if (!tcmu_dev_flags_test(dev, TCMU_DEV_FLAG_REOPENING)) {
+			ret = -EINVAL;
+			goto close;
+		}
+
+		/*
+		 * If we are here this should be in reopen path,
+		 * then we should also update the device size in
+		 * kernel.
+		 */
+		tcmu_dev_info(dev,
+			      "device size and backing size disagree:device %lld backing %lld\n",
+			      dev_size, (long long) st.st_size);
+
+		ret = tcmu_update_num_lbas(dev, st.st_size);
+		if (!ret) {
+			ret = tcmu_set_device_size(dev, false);
+			if (ret) {
+				/* Rolling back */
+				tcmu_update_num_lbas(dev, dev_size);
+				goto close;
```

Comment:
Let's just put the update_num_lbas, set_dev_size and roll back error handling, in a function like tcmu_dev_update_size().

Also in that new function drop the SIZE_CHANGES bits and just add a call to tcmur_set_pending_ua().

Reply:
Agree and will update it.

Reply:
BTW, How to deal with this issue ? 

Linking C executable consumer
libtcmu.so.2: undefined reference to `tcmur_set_pending_ua'
collect2: error: ld returned 1 exit status
make[2]: *** [consumer] Error 1
make[1]: *** [CMakeFiles/consumer.dir/all] Error 2
make: *** [all] Error 2

The tcmur_set_pending_ua is in tcmur_cmd_handler.c and we will use it in libtcmu.so.

Reply:
You can just put the tcmu_dev_update_size in tcmur_device.c then add it to main-syms.txt for now.

Again, when the new lib is made we will sort all this out in a proper lib.